tests/mtouch/InlinerTest.cs

```diff
@@ -0,0 +1,157 @@
+﻿using NUnit.Framework;
+using System;
+using System.IO;
+using System.Linq;
+using System.Collections.Generic;
+using Mono.Cecil;
+using Mono.Cecil.Cil;
+using Xamarin.Tests;
+
+namespace Xamarin.Linker {
+
+	public static partial class Extensions {
+
+		// note: direct check, no inheritance
+		public static bool Is (this TypeReference type, string @namespace, string name)
+		{
+			return ((type != null) && (type.Name == name) && (type.Namespace == @namespace));
+		}
+	}
+
+	public abstract class InlinerTest {
+
+		// note: not every candidate should be handled by the linker 
+		// most of them are _naturally_ eliminated by not being used/marked
+		static bool DisplayCandidates = false;
+
+		protected abstract string Assembly { get; }
+
+		AssemblyDefinition assembly;
+
+		protected virtual AssemblyDefinition AssemblyDefinition {
+			get {
+				if (assembly == null)
+					assembly = AssemblyDefinition.ReadAssembly (Assembly);
+				return assembly;
+			}
+		}
+
+		protected string ListMethods (HashSet<string> h)
+		{
+			string result = Path.GetFileName (Assembly);
+			if (h.Count == 0)
+				return $" {result}: success";
+			return result + "\n" + h.Aggregate ((arg1, arg2) => arg1 + "\n" + arg2);
+		}
+
+		bool IsCandidateForInlining (MethodDefinition m)
+		{
+			// candidates must be methods
+			if (m.IsConstructor)
+				return false;
+			// must have a body (IL)
+			if (!m.HasBody)
+				return false;
+			// must be static or not virtual (can't be overrriden)
+			if (!m.IsStatic && m.IsVirtual)
+				return false;
+			// the body must not have exception handlers or variables
+			var b = m.Body;
+			return !(b.HasExceptionHandlers || b.HasVariables || b.InitLocals);
+		}
+
+		protected void NoParameterReturnOnlyConstant (Code code, HashSet<string> list)
+		{
+			if (DisplayCandidates)
+				Console.WriteLine ($"### NoParameterReturnOnlyConstant {code}: {Path.GetFileName (Assembly)}");
+
+			var ad = AssemblyDefinition.ReadAssembly (Assembly);
+			foreach (var t in ad.MainModule.Types) {
+				if (!t.HasMethods)
+					continue;
+				foreach (var m in t.Methods) {
+					if (!IsCandidateForInlining (m))
+						continue;
+					if (m.HasParameters)
+						continue;
+					var b = m.Body;
+					if (b.Instructions.Count != 2)
+						continue;
+					var ins = b.Instructions [0];
+					if (code == ins.OpCode.Code) {
+						var s = m.ToString ();
+						list.Remove (s);
```

Comment:
Ah, you're right about HashSet. However `ListMethods` will show if a method we already inline changes (if it's not inlineable anymore), but it will not show if a method becomes inlineable (either because it's new, or its implementation changed) - but I guess that wasn't your intention.